### ISSUE TITLE
The `get` function of the sparse matrix `csr` with stream does not give excepted result.

### ISSUE THREAD

**Issue by szsdk** [cupy/cupy#4136]
* Conditions
CuPy Version          : 8.0.0
CUDA Root             : /usr/local/cuda-11.0
CUDA Build Version    : 11000
CUDA Driver Version   : 11000
CUDA Runtime Version  : 11000
cuBLAS Version        : 11200
cuFFT Version         : 10201
cuRAND Version        : 10201
cuSOLVER Version      : (10, 6, 0)
cuSPARSE Version      : 11101
NVRTC Version         : (11, 0)
Thrust Version        : 100909
CUB Build Version     : 100909
cuDNN Build Version   : None
cuDNN Version         : None
NCCL Build Version    : 2708
NCCL Runtime Version  : 2708
cuTENSOR Version      : None
* Code to reproduce
```python
import numpy as np
import cupy as cp
from cupy import cuda
from cupyx.scipy.sparse import csr_matrix as cu_csr
from scipy.sparse import csr_matrix

dense_h = np.random.rand(20, 10000)
dense_h[dense_h < 0.95] = 0
sparse_h = csr_matrix(dense_h)
dense_d = cp.array(dense_h)

asp_d = cu_csr(dense_d)
asp_h = asp_d.get()

assert np.all(asp_h.A == sparse_h.A)

s = cuda.Stream(non_blocking=True)
# Swapping next two lines does not help
with s:
    asp_d = cu_csr(dense_d)
    asp_h = asp_d.get(stream=s)
    s.synchronize()

assert np.all(asp_h.A == sparse_h.A)
```

**Comment by szsdk** [comment]
I solved this problem. It turns out that cupy sparse does not set stream at all. Now, I need manually do this. Hope in the futrue, a smarter API or more clear documentation will be applied.
```python
import numpy as np
import cupy as cp
import time
from cupy import cuda

from cupyx.scipy.sparse import csr_matrix as cu_csr
from scipy.sparse import csr_matrix

dense_h = np.random.rand(20, 10000)
dense_h[dense_h < 0.95] = 0
sparse_h = csr_matrix(dense_h)
dense_d = cp.array(dense_h)

asp_d = cu_csr(dense_d)
asp_h = asp_d.get()

assert np.all(asp_h.A == sparse_h.A)

# ############
from cupy.cuda import device as _device
from cupy_backends.cuda.libs import cusparse
def set_cusparse_stream(s):
    cusparse.setStream(_device.get_cusparse_handle(), s.ptr)
# ############

s = cuda.Stream(non_blocking=True)
with s:
    set_cusparse_stream(s) # commenting this line causes wrong copy.
    asp_d = cu_csr(dense_d)
    asp_h = asp_d.get(stream=s)
    s.synchronize()

assert np.all(asp_h.A == sparse_h.A)
```

**Comment by takagi** [comment]
Thanks for reporting. CuPy sparse is basically designed to automatically use the current stream if it is set:

https://github.com/cupy/cupy/blob/a8dfcd66d89c8e66a60e4b7272f95a15c26fc907/cupy_backends/cuda/libs/cusparse.pyx#L1556-L1560

There may be something wrong. Let me have further investigation.

**Comment by takagi** [comment]
Would you try https://github.com/cupy/cupy/pull/4173 to check if it fixes the issue, if possible? @szsdk

**Comment by szsdk** [comment]
I'd love to. It seems that I need to compile cupy by myself. Could you kindly point out how to compile it with multiple cores? The compiling is a bit slow now.

**Comment by szsdk** [comment]
I have tested. The example code I posted passed in @takagi 's patch.

**Comment by takagi** [comment]
Thanks a lot! The fix is now included in the next stable minor v8.1, which is planed to release this Thursday.
